blocks/search.py

```diff
@@ -0,0 +1,370 @@
+from abc import ABCMeta, abstractmethod
+from collections import OrderedDict
+from six.moves import range
+
+import numpy
+
+from theano import config
+from theano import function
+from theano import tensor
+
+from blocks.filter import VariableFilter
+from blocks.graph import ComputationGraph
+
+floatX = config.floatX
+
+
+class Search(object):
+    """Abstract search class.
+
+    Parameters
+    ----------
+    sequence_generator : sequence generator
+        Sequence generator to be used
+
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, sequence_generator):
+        self.generator = sequence_generator
+        self.compiled = False
+
+    @abstractmethod
+    def compile(self):
+        self.compiled = True
+
+    @abstractmethod
+    def search(self, **kwargs):
+        r"""Performs search.
+
+        Parameters
+        ----------
+        \*\*kwargs :
+            Arguments needed by sequence generator
+
+        Returns
+        -------
+        Generated sequences
+        """
+        if not self.compiled:
+            self.compile()
+
+
+class GreedySearch(Search):
+
+    def __init__(self, sequence_generator):
+        super(GreedySearch, self).__init__(sequence_generator)
+
+
+class BeamSearch(Search):
```

Comment:
Greedy search is just beam search with beam width equal to one, so I do not see why we need this one and the base class.